<a href="https://colab.research.google.com/github/Gathoni005/ML_projects/blob/main/customer_churn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
customer_churn_df=pd.read_csv('/content/drive/MyDrive/WA_Fn-UseC_-Telco-Customer-Churn.csv')

In [4]:
customer_churn_df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [5]:
customer_churn_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


according to the data collected most of the data are not in (int) datatype and they can't work if they are objects


In [6]:
customer_churn_df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

there are no null values

In [7]:
customer_churn_df['customerID']=customer_churn_df['customerID'].str.split('-').str[0].astype(int)

transforming the customerid column to an integer datatype and removig the word part by replacing it with an empty


In [8]:
customer_churn_df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0        7590  Female              0     Yes         No       1           No   
1        5575    Male              0      No         No      34          Yes   
2        3668    Male              0      No         No       2          Yes   
3        7795    Male              0      No         No      45           No   
4        9237  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [9]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
categorical_cols = ['gender', 'Partner', 'Dependents', 'PhoneService',
                    'MultipleLines','InternetService', 'OnlineSecurity','OnlineBackup', 'DeviceProtection',
                    'TechSupport', 'StreamingTV', 'StreamingMovies','Contract',
                    'PaperlessBilling','PaymentMethod', 'Churn']
for col in categorical_cols:
  customer_churn_df[col] = le.fit_transform(customer_churn_df[col])

using the label encoder method from sklearn.preprocessing to convert categorical values in the dataframe to numerical values

In [10]:
customer_churn_df.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0        7590       0              0        1           0       1   
1        5575       1              0        0           0      34   
2        3668       1              0        0           0       2   
3        7795       1              0        0           0      45   
4        9237       0              0        0           0       2   

   PhoneService  MultipleLines  InternetService  OnlineSecurity  ...  \
0             0              1                0               0  ...   
1             1              0                0               2  ...   
2             1              0                0               2  ...   
3             0              1                0               2  ...   
4             1              0                1               0  ...   

   DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
0                 0            0            0                0         0   
1                 2            0            0                0         1   
2                 0            0            0                0         0   
3                 2            2            0                0         1   
4                 0            0            0                0         0   

   PaperlessBilling  PaymentMethod  MonthlyCharges  TotalCharges Churn  
0                 1              2           29.85         29.85     0  
1                 0              3           56.95        1889.5     0  
2                 1              3           53.85        108.15     1  
3                 0              0           42.30       1840.75     0  
4                 1              2           70.70        151.65     1  

[5 rows x 21 columns]

In [ ]:
print(customer_churn_df_encoded)

     customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0          7590       0              0        1           0       1   
1          5575       1              0        0           0      34   
2          3668       1              0        0           0       2   
3          7795       1              0        0           0      45   
4          9237       0              0        0           0       2   
...         ...     ...            ...      ...         ...     ...   
7038       6840       1              0        1           1      24   
7039       2234       0              0        1           1      72   
7040       4801       0              0        1           1      11   
7041       8361       1              1        1           0       4   
7042       3186       1              0        0           0      66   

      PhoneService  MultipleLines  OnlineSecurity OnlineBackup  ...  Contract  \
0                0              1               0          Yes  ..

In [11]:
customer_churn_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   int64  
 1   gender            7043 non-null   int64  
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   int64  
 4   Dependents        7043 non-null   int64  
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   int64  
 7   MultipleLines     7043 non-null   int64  
 8   InternetService   7043 non-null   int64  
 9   OnlineSecurity    7043 non-null   int64  
 10  OnlineBackup      7043 non-null   int64  
 11  DeviceProtection  7043 non-null   int64  
 12  TechSupport       7043 non-null   int64  
 13  StreamingTV       7043 non-null   int64  
 14  StreamingMovies   7043 non-null   int64  
 15  Contract          7043 non-null   int64  
 16  PaperlessBilling  7043 non-null   int64  


only the total charges have not changed to numerical values according to the information collected

In [12]:
missing_total_charges = customer_churn_df['TotalCharges'].isnull().sum()
print(f"Missing values in 'TotalCharges': {missing_total_charges}")

Missing values in 'TotalCharges': 0


In [13]:
blank_values = (customer_churn_df['TotalCharges'].str.strip() == '').sum()
print(f"Blank or empty string values in 'TotalCharges': {blank_values}")

Blank or empty string values in 'TotalCharges': 11


In [17]:
import numpy as np
customer_churn_df['TotalCharges'] = customer_churn_df['TotalCharges'].replace(r'^\s*$', np.nan, regex=True)
customer_churn_df['TotalCharges'] = pd.to_numeric(customer_churn_df['TotalCharges'], errors='coerce')
customer_churn_df['TotalCharges'].fillna(0, inplace=True)


<ipython-input-17-485ae2de3b29>:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  customer_churn_df['TotalCharges'].fillna(0, inplace=True)


repacing empty or invalid values in the total charges column with nun converting the column to a numeric type and replacing nan values with 0

In [18]:
customer_churn_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   int64  
 1   gender            7043 non-null   int64  
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   int64  
 4   Dependents        7043 non-null   int64  
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   int64  
 7   MultipleLines     7043 non-null   int64  
 8   InternetService   7043 non-null   int64  
 9   OnlineSecurity    7043 non-null   int64  
 10  OnlineBackup      7043 non-null   int64  
 11  DeviceProtection  7043 non-null   int64  
 12  TechSupport       7043 non-null   int64  
 13  StreamingTV       7043 non-null   int64  
 14  StreamingMovies   7043 non-null   int64  
 15  Contract          7043 non-null   int64  
 16  PaperlessBilling  7043 non-null   int64  


In [19]:
customer_churn_df.corr()

customerID    gender  SeniorCitizen   Partner  Dependents  \
customerID          1.000000  0.006073      -0.002197 -0.026779   -0.012816   
gender              0.006073  1.000000      -0.001874 -0.001808    0.010517   
SeniorCitizen      -0.002197 -0.001874       1.000000  0.016479   -0.211185   
Partner            -0.026779 -0.001808       0.016479  1.000000    0.452676   
Dependents         -0.012816  0.010517      -0.211185  0.452676    1.000000   
tenure              0.007805  0.005106       0.016567  0.379697    0.159712   
PhoneService       -0.006252 -0.006488       0.008576  0.017706   -0.001762   
MultipleLines       0.004366 -0.006739       0.146185  0.142410   -0.024991   
InternetService    -0.012230 -0.000863      -0.032310  0.000891    0.044590   
OnlineSecurity      0.013174 -0.015017      -0.128221  0.150828    0.152166   
OnlineBackup       -0.003351 -0.012057      -0.013632  0.153130    0.091015   
DeviceProtection   -0.007084  0.000549      -0.021398  0.166330    0.080537   
TechSupport         0.001106 -0.006825      -0.151268  0.126733    0.133524   
StreamingTV        -0.007995 -0.006421       0.030776  0.137341    0.046885   
StreamingMovies    -0.016800 -0.008743       0.047266  0.129574    0.021321   
Contract            0.014765  0.000126      -0.142554  0.294806    0.243187   
PaperlessBilling   -0.001904 -0.011754       0.156530 -0.014877   -0.111377   
PaymentMethod       0.011835  0.017352      -0.038551 -0.154798   -0.040292   
MonthlyCharges     -0.003995 -0.014569       0.220173  0.096848   -0.113890   
TotalCharges        0.000185 -0.000080       0.103006  0.317504    0.062078   
Churn              -0.017385 -0.008612       0.150889 -0.150448   -0.164221   

                    tenure  PhoneService  MultipleLines  InternetService  \
customerID        0.007805     -0.006252       0.004366        -0.012230   
gender            0.005106     -0.006488      -0.006739        -0.000863   
SeniorCitizen     0.016567      0.008576       0.146185        -0.032310   
Partner           0.379697      0.017706       0.142410         0.000891   
Dependents        0.159712     -0.001762      -0.024991         0.044590   
tenure            1.000000      0.008448       0.343032        -0.030359   
PhoneService      0.008448      1.000000      -0.020538         0.387436   
MultipleLines     0.343032     -0.020538       1.000000        -0.109216   
InternetService  -0.030359      0.387436      -0.109216         1.000000   
OnlineSecurity    0.325468     -0.015198       0.007141        -0.028416   
OnlineBackup      0.370876      0.024105       0.117327         0.036138   
DeviceProtection  0.371105      0.003727       0.122318         0.044944   
TechSupport       0.322942     -0.019158       0.011466        -0.026047   
StreamingTV       0.289373      0.055353       0.175059         0.107417   
StreamingMovies   0.296866      0.043870       0.180957         0.098350   
Contract          0.671607      0.002247       0.110842         0.099721   
PaperlessBilling  0.006152      0.016505       0.165146        -0.138625   
PaymentMethod    -0.370436     -0.004184      -0.176793         0.086140   
MonthlyCharges    0.247900      0.247398       0.433576        -0.323260   
TotalCharges      0.826178      0.113214       0.452577        -0.175755   
Churn            -0.352229      0.011942       0.038037        -0.047291   

                  OnlineSecurity  ...  DeviceProtection  TechSupport  \
customerID              0.013174  ...         -0.007084     0.001106   
gender                 -0.015017  ...          0.000549    -0.006825   
SeniorCitizen          -0.128221  ...         -0.021398    -0.151268   
Partner                 0.150828  ...          0.166330     0.126733   
Dependents              0.152166  ...          0.080537     0.133524   
tenure                  0.325468  ...          0.371105     0.322942   
PhoneService           -0.015198  ...          0.003727    -0.019158   
MultipleLines           0.0071

according to the correlation customerid column has the lowest correlation

In [20]:
customer_churn_df=customer_churn_df.drop(columns=["customerID"])

dropping the customerid column

SPLITTING OF THE DATASET


In [21]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

the code snippet from scikit learn library for logistic regression

In [22]:
x=customer_churn_df.drop(columns=["Churn"])
y=customer_churn_df["Churn"]
print(f"Features (x) shape: {x.shape}")
print(f"Target (y) shape: {y.shape}")

Features (x) shape: (7043, 19)
Target (y) shape: (7043,)


features set x contains 7043 rows and 19columns
target variable y contains 7043 entries

In [23]:
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42, stratify=y)

splitting the datasets into training and testing sets

In [24]:
model = LogisticRegression(max_iter=1000)

In [26]:
model.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=1000)

In [27]:
y_pred = model.predict(X_test)

In [ ]:
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))
print(f"Accuracy Score: {accuracy_score(y_test, y_pred)}")

Confusion Matrix:
[[917 118]
 [166 208]]

Classification Report:
              precision    recall  f1-score   support

           0       0.85      0.89      0.87      1035
           1       0.64      0.56      0.59       374

    accuracy                           0.80      1409
   macro avg       0.74      0.72      0.73      1409
weighted avg       0.79      0.80      0.79      1409

Accuracy Score: 0.7984386089425124


approximately 80% of all predictions was correct